controllers/cps_controller_helper.go

```diff
@@ -0,0 +1,293 @@
+			BindAddress: ":8083",
+		},
+		LeaderElection:   r.EnableLeaderElection,
+		LeaderElectionID: "5ff985ccc.constraintstatuspod.gatekeeper.sh",
+		Cache: cacheRuntime.Options{
+			ByObject: map[client.Object]cacheRuntime.ByObject{
+				&gkv1beta1.ConstraintPodStatus{}: {
+					Transform: func(obj interface{}) (interface{}, error) {
+						constraintStatus := obj.(*gkv1beta1.ConstraintPodStatus)
+						// Only cache fields that are utilized by the controllers.
+						guttedObj := &gkv1beta1.ConstraintPodStatus{
+							TypeMeta: constraintStatus.TypeMeta,
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      constraintStatus.Name,
+								Labels:    constraintStatus.Labels,
+								Namespace: constraintStatus.Namespace,
+							},
+							Status: gkv1beta1.ConstraintPodStatusStatus{
+								ObservedGeneration: constraintStatus.Status.ObservedGeneration,
+								Operations:         constraintStatus.Status.Operations,
+							},
+						}
+
+						return guttedObj, nil
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		setupLog.Error(err, "Failed to setup NewManager for ConstraintPodStatus contoller")
+
+		return err
+	}
+
+	constraintToSyncOnly := r.getConstraintToSyncOnly(mainCtx)
+
+	if err := (&ConstraintPodStatusReconciler{
+		Scheme:               r.Scheme,
+		Client:               cpsMgr.GetClient(),
+		DynamicClient:        r.DynamicClient,
+		Log:                  ctrl.Log.WithName("ConstraintPodStatus"),
+		Namespace:            r.Namespace,
+		ConstraintToSyncOnly: constraintToSyncOnly,
+		DiscoveryStorage:     r.DiscoveryStorage,
+	}).SetupWithManager(cpsMgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "ConstraintPodStatus")
+
+		return err
+	}
+
+	r.isCPSCtrlRunning = true
+
+	// Use another go routine for the ConstraintPodStatus controller
+	go func() {
+		err := cpsMgr.Start(cpsCtrlCtx)
+		if err != nil {
+			setupLog.Error(err, "A problem running ConstraintPodStatus manager. Triggering a reconcile to restart it.")
+		}
+
+		defer r.cpsCtrlCtxCancel()
+
+		r.cpsCtrlCtxCancel = nil
+		r.isCPSCtrlRunning = false
+
+		// In case it is not an error and a child context is cancelled
+		// because the auditFromCache changed from Automatic,
+		// sending this channel avoids encountering a race condition.
+		// If the error happens when cpsMgr start, it will retry to start cpsMgr
+		r.ManualReconcileTrigger <- event.GenericEvent{
+			Object: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": v1alpha1.GroupVersion.String(),
+					"kind":       "Gatekeeper",
+					"metadata": map[string]interface{}{
+						"name": defaultGatekeeperCrName,
+					},
+				},
+			},
+		}
+	}()
+
+	return nil
+}
+
+func (r *GatekeeperReconciler) getConstraintToSyncOnly(mainCtx context.Context) map[string][]gkv1alpha1.SyncOnlyEntry {
+	cpsList := &v1beta1.ConstraintPodStatusList{}
+
+	err := r.Client.List(mainCtx, cpsList, &client.ListOptions{})
+	if err != nil {
+		return nil
```

Comment:
I think in `ConstraintPodStatusReconciler` you always assume a non-nil `constraintToSyncOnly` right? You can read from a nil map but you can't write to it. So make sure this is safe, otherwise, this method should always return an empty map instead of nil.

Reply:
so `cpsList.Items` is 0 then here 
```
	if err := (&ConstraintPodStatusReconciler{
		Scheme:               r.Scheme,
		Client:               cpsMgr.GetClient(),
		DynamicClient:        r.DynamicClient,
		Log:                  ctrl.Log.WithName("ConstraintPodStatus"),
		Namespace:            r.Namespace,
		ConstraintToSyncOnly: constraintToSyncOnly,
		DiscoveryStorage:     r.DiscoveryStorage,
	}).SetupWithManager(cpsMgr); err != nil {
		
```
ConstraintToSyncOnly is assigned as` map[string][]gkv1alpha1.SyncOnlyEntry 
`

Reply:
gotcha